src/bin/permonsvmfile.c

```diff
@@ -114,6 +120,45 @@ PetscErrorCode SVMRunBinaryClassification()
     TRY( SVMTest(svm) );
   }
 
+  /* Save results */
+  if (training_result_file_set) {
+    TRY( GetFilenameExtension(training_result_file,&extension) );
+    TRY( PetscStrcmp(extension,h5,&ishdf5) );
+    TRY( PetscStrcmp(extension,bin,&isbinary) );
+    if (ishdf5) {
+#if defined(PETSC_HAVE_HDF5)
+      TRY( PetscViewerHDF5Open(PETSC_COMM_WORLD,training_result_file,FILE_MODE_WRITE,&viewer) );
+#else
+      SETERRQ(PETSC_COMM_WORLD,PETSC_ERR_SUP,"PETSc is not configured with HDF5");
+#endif
+    } else if (isbinary) {
+      TRY( PetscViewerBinaryOpen(PETSC_COMM_WORLD,training_result_file,FILE_MODE_WRITE,&viewer) );
+    } else {
+      SETERRQ1(PETSC_COMM_WORLD,PETSC_ERR_SUP,"File type %s not supported",extension);
+    }
+    TRY( SVMViewTrainingPredictions(svm,viewer) );
+    TRY( PetscViewerDestroy(&viewer) );
+  }
+
+  if (test_result_file_set) {
+    TRY( GetFilenameExtension(test_result_file,&extension) );
+    TRY( PetscStrcmp(extension,h5,&ishdf5) );
+    TRY( PetscStrcmp(extension,bin,&isbinary) );
+    if (ishdf5) {
+#if defined(PETSC_HAVE_HDF5)
+      TRY( PetscViewerHDF5Open(PETSC_COMM_WORLD,test_result_file,FILE_MODE_WRITE,&viewer) );
+#else
+      SETERRQ(PETSC_COMM_WORLD,PETSC_ERR_SUP,"PETSc is not configured with HDF5");
+#endif
+    } else if (isbinary) {
+      TRY( PetscViewerBinaryOpen(PETSC_COMM_WORLD,test_result_file,FILE_MODE_WRITE,&viewer) );
+    } else {
+      SETERRQ1(PETSC_COMM_WORLD,PETSC_ERR_SUP,"File type %s not supported",extension);
```

Comment:
I'd just default to binary file if the extension is not *.h5. So that -f_test_predictions out works.

Reply:
Well. It makes sense when PermonSVM is only one tool in an ML stack. However, in practical applications, where PermonSVM could be incorporated (into a model layer), there are plenty of other ML tools. Each one should use its own proprietary format for storing results. We can show in this tutorial that the PETSc binary file and hdf5 are currently supported.

Reply:
Last argument is a good one. Changed.